src/replication.c

```diff
@@ -2386,6 +2481,353 @@ char *sendCommandArgv(connection *conn, int argc, char **argv, size_t *argv_lens
+        }
+        abortRdbConnectionSync(1);
+        return C_ERR;
+    }
+    o->used += nread;
+    incrReadsProcessed(nread);
+    return read - nread;
+}
+
+int isReplicaBufferLimitReached(void) {
+    return server.pending_repl_data.len > server.client_obuf_limits[1].hard_limit_bytes;
+}
+
+/* Read handler for buffering incoming repl data during RDB download/loading. */
+void bufferReplData(connection *conn) {
+    client *c = connGetPrivateData(conn);
+    size_t readlen = PROTO_IOBUF_LEN;
+    int read = 0;
+    
+    while (readlen > 0) {
+        listNode *ln = listLast(server.pending_repl_data.blocks);
+        replDataBufBlock *tail = ln ? listNodeValue(ln) : NULL;
+
+        /* Append to tail string when possible */
+        if (tail && tail->used < tail->size) {
+            size_t avail = tail->size - tail->used;
+            read = min(readlen, avail);
+            readlen -= read;
+            read = readIntoReplDataBlock(conn, tail, read);
+        }
+        if (readlen && read == 0) {
+            if (isReplicaBufferLimitReached()) {
+                serverLog(LL_DEBUG, "Replication buffer limit reached, stopping buffering");
+                break;
+            }
+            /* Create a new node */
+            tail = zmalloc(sizeof(replDataBufBlock));
+            tail->size = PROTO_IOBUF_LEN;
+            tail->used = 0;
+            listAddNodeTail(server.pending_repl_data.blocks, tail);
+            server.pending_repl_data.len += tail->size;
+
+            read = min(readlen, tail->size);
+            readlen -= read;
+            read = readIntoReplDataBlock(conn, tail, read);            
+        }
+        if (read > 0) {
+            /* Stop reading in case we read less than we anticipated */
+            break;
+        }
+        if (read == C_ERR) {
+            return;
+        }
+    }    
+    c->lastinteraction = server.unixtime;  
+}
+
+/* Streams accumulated replication data into the database while freeing read nodes */
+void streamReplDataBufToDb(client *c) {
+    serverAssert(c->flags & CLIENT_MASTER);
+    blockingOperationStarts();
+    size_t offset = 0;
+    listNode *cur = NULL;
+
+    while ((cur = listFirst(server.pending_repl_data.blocks))) {
+        /* Read and process repl data block */
+        replDataBufBlock *o = listNodeValue(cur);
+        c->querybuf = sdscatlen(c->querybuf, o->buf, o->used);
+        c->read_reploff += o->used;
+        processInputBuffer(c);
+        offset += o->used;
+        replStreamProgressCallback(offset, o->used);
+
+        listDelNode(server.pending_repl_data.blocks, cur);
+    } 
+
+    blockingOperationEnds();
+}
+
+/* There are two scenarios in which this method can be called during rdb-channel-sync:
+ * 1. Main connection successfully established psync with primary.
+ * 2. Rdb connection done loading rdb. 
+ * Each time this method is invoked, we check whether the other connection has completed
+ * his part and act accordingly */
```

Comment:
```suggestion
/* There are two scenarios in which this method can be called during rdb-channel-sync:
 * 1. Main connection successfully established psync with master.
 * 2. Rdb connection done loading rdb. 
 *
 * The 'conn' argument is either server.repl_transfer_s or server.repl_full_sync_s.
 * Each time this method is invoked, we check whether the other connection has completed
 * its part and act accordingly. */
```
don't forget to update the variable names if renamed.

Reply:
p.s. i feel the function name doesn't accurately describe what it does or when it's called or what purpose it fills (the comment does).
maybe repConnectionDone? or something that indicates it's the final state of one of the connections...
do you have a better idea?

Reply:
function does need to be renamed, I think `completeRDBChannelSync` may be a good fit.